[MISSION1] 자바 야구게임 구현

JakePark929 opened the pull request:
자바 야구게임 Mission 1번 진행하였습니다.
아래는 각 클래스에 대한 간략한 설명입니다.

- `Baseball`: 사용자의 Input 값을 Validation 하고 저장합니다.
- `Baseballs`: 3자리 숫자리스트를 Validation 하여 보관하는 도메인 입니다.
- `BaseballFactory`: 숫자리스트 생성을 담당하는 클래스 입니다.
- `Computer` : 랜덤한 숫자리스트를 생성을 요청합니다.
- `BaseballGame` : 게임의 전반적인 진행 상태를 보관합니다.
- `GameManager` : 게임 진행을 도와주는 클래스 입니다.
- `Result` : 예측값 입력시 결과 상태를 보관하고 출력하는 클래스 입니다.
- `Input` : 사용자 입력 인터페이스를 담당합니다.
- `Output` : 진행상황 출력 인터페이스를 담당합니다.

바쁘시겠지만 코드 리뷰 잘 부탁드리겠습니다. :)

JakePark929 committed:
[MISSION1] 자바 야구게임 구현

GameManager 와 BaseballFactory 클래스 등을 이용한 객체지향 야구게임 구현

- `Baseball`: 사용자의 Input 값을 Validation 하고 저장합니다.
- `Baseballs`: 3자리 숫자리스트를 Validation 하여 보관하는 도메인 입니다.
- `BaseballFactory`: 숫자리스트 생성을 담당하는 클래스 입니다.
- `Computer` : 랜덤한 숫자리스트를 생성을 요청합니다.
- `BaseballGame` : 게임의 전반적인 진행 상태를 보관합니다.
- `GameManager` : 게임 진행을 도와주는 클래스 입니다.
- `Result` : 예측값 입력시 결과 상태를 보관하고 출력하는 클래스 입니다.
- `Input` : 사용자 입력 인터페이스를 담당합니다.
- `Output` : 진행상황 출력 인터페이스를 담당합니다.

donsonioc2010 commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private Baseballs receiveUserBaseballs() {
+        try {
+            return BaseballFactory.createBaseballs(Input.receiveNumberInput());
+        } catch (IllegalArgumentException e) {
+            Output.printExceptionMessage(e.getMessage());
+
+            return receiveUserBaseballs();
+        }
```
오 이런 Catch문은 생각을 못해봤는데 영감을 얻어갑니다. 😃

donsonioc2010 commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,43 @@
+
+        do {
+            playBaseballOnce();
+        } while(!isCorrectAnswer());
```
제가 많이 부족해서인지, 중언님코드를 재밌게 읽다가 딱 한곳 여기만 눈에 보였습니다 ㅎㅎ :)

여기에서 do-while문을 꼭 사용해야 했을까?라는 생각이 많이 들더라구요. 
이유가 과거에 클린코드를 읽으면서 do-while문을 최대한 자제하라라는 문구를 흘려봤었고, 회사에서도 근무하면서 do-while문장들의 경우 한번더 생각해야 하다보니 불편함을 많이 느꼈던게 이유인것 같습니다.

한번 같이 생각해보면 좋지 않을까 합니다. 👍

donsonioc2010 commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private boolean isCorrectAnswer() {
+        return userBaseballs.equals(computerBaseballs);
+    }
```
이 부분이 위에서 기록한 것처럼 do-while문을 쓸수 밖에 없었던 이유라고 생각이 듭니다.

저의 경우에는 다음의 가정을 해볼것 같았는데요. 

> `playBaseball`메소드를 실행 중인 상황에서, `computerBaseballs` 변수나  `userBaseballs`둘중 어떤 변수가 무조건 `null`이 절대 나오지 않는 변수일까? 를 가정해 두지 않을 까 싶습니다.

JakePark929 replied:
이 부분 사실 가정하는 것이 무슨 말씀인지 잘 이해가 안 가서 우선 리뷰 반영했는데 맞게 한 건지 모르겠네요!
확인 부탁 드려요~

donsonioc2010 replied:
저도 다시 읽어보니 약간 혼동이 있을 수 있겠다는게 느껴지네요. (다시 읽어보다보니 저도 헷갈리네요)

제가 말하고자 했던 내용은, 위에서 do~while을 사용할 수밖에 없었고, 해당 이유는 코멘트가 달려있는 `isCorrectAnswer()` 메소드 떄문이겠다 라고 생각을 하였습니다.

이유는 해당 메소드에서 `userBaseballs.equals(computerBaseballs)` 다음 조건으로 사용하게 될경우 NPE가 발생할 수 있기 때문에 do~while을 사용하셨겠구나 라고 인식을 하였습니다.

그러면 여기서 `computerBaseballs`과 `userBaseballs`를 봤을때, 과연 `computerBaseballs`변수는 Null이라는 케이스가 발생할 수 있을까? 반대로 `userBaseballs`라는 변수는 Null이라는 케이스가 발생 할 수 있을까? 라는 해당 두 케이스를 생각해보라는 의미였습니다. :)

위의 문제를 생각해보면 왠지 위의 do~while에서도 while문으로 바로 시작하는 방법으로 변경해 볼수 있지 않을까? 하는 의미에서 남겼던 코멘트입니다.

JakePark929 replied:
그런 말씀이었군요! 이해 했습니다.
코드는 while 문으로 바꿨습니다!
좋은 의견 감사합니다 ㅎㅎ

donsonioc2010 reviewed (approved):
코드 재밌게 읽었습니다 :) 

정말 딱히 할말이 없었습니다 👍 👍 👍  

저도 배우는 입장이다 보니, 제가 틀린거다 싶으면 같이 의견 남겨주시면 감사하겠습니다 :)

동현님도 보셔야하니, 리뷰는 Comment로 처리해두도록 하겠습니다 😃

birdieHyun commented on src/test/java/baseball/domain/BaseballsTest.java:
```diff
@@ -0,0 +1,54 @@
+        Baseballs ask3 = BaseballFactory.createBaseballs("261");
+        Baseballs ask4 = BaseballFactory.createBaseballs("231");
+
+        assertEquals(0, answer.countBall(ask1), "낫싱");
+        assertEquals(1, answer.countBall(ask2), "1 볼");
+        assertEquals(2, answer.countBall(ask3), "2 볼");
+        assertEquals(3, answer.countBall(ask4), "3 볼");
```
비교적 간단한 테스트라서 큰 문제는 없겠지만 하나의 @Test 안에 여러번의 assert를 할 경우 assertAll() 로 묶어주는건 어떨까요?

JakePark929 replied:
위 쪽에 assertAll() 쓰다가 이 부분 빼먹었네요! 
반영완료했습니다!

birdieHyun commented on src/main/java/baseball/service/Result.java:
```diff
@@ -0,0 +1,34 @@
+    private final int strikeNumber;
+    private final int ballNumber;
+
+    private static final int COUNT_ZERO = 0;
```
상수의 경우 클래스의 맨 위에 배치하여 가독성을 높이는건 어떨까요?

birdieHyun reviewed (approved):
고생많으셨습니다 :) 코드를 잘 작성하셔서 리뷰해드릴게 별로 없네요

JakePark929 committed:
refactor: do-while 문 지양 외 코드 리뷰 반영

do-while 문을 while 문으로 변경
Result 클래스 상수 상단 위치, test 중 반복된 assert 를 assertAll 로 변경

JakePark929 commented:
좋은 리뷰 감사합니다!
리뷰 내용 반영한 거 코멘트 받은 후에 머지 하겠습니다.👍
